File: neetcode150/13_1d_dp/1143_longest_common_subsequence.py
"""
=====================================================================
LeetCode 1143 : Longest Common Subsequence                      (Medium)
https://leetcode.com/problems/longest-common-subsequence/
Category   : 1-D dynamic programming
---------------------------------------------------------------------
PROBLEM
    Length of the longest common subsequence of two strings.
---------------------------------------------------------------------
INTUITION
    dp[i][j] = LCS of the first i chars of text1 and first j of
    text2. Equal chars → 1 + dp[i-1][j-1]; else the max of dropping
    either tail. Rolling rows cut memory to O(min(m,n)).
---------------------------------------------------------------------
APPROACH 1 · Full DP table
    Standard 2D recurrence, easy to read and verify.
APPROACH 2 · Rolling two rows (memory-optimized)
    prev/cur rows; identical recurrence, O(n) memory.
---------------------------------------------------------------------
DEEP DIVE · Longest-path interpretation
    State (i,j) on a grid; a diagonal step (match) earns +1, right/
    down steps are free — dp is a longest path. Optimal substructure
    holds because a prefix of an optimal LCS is optimal for the
    corresponding prefix pair. 1000×1000 strings = 1M table cells
    (~8 MB) vs 8 KB for the rolling variant.
---------------------------------------------------------------------
TIME COMPLEXITY : O(m·n)
MEMORY COMPLEXITY: O(m·n) or O(n) rolling.
=====================================================================
"""


class Solution_1:
    def longestCommonSubsequence(self, text1, text2):
        m, n = len(text1), len(text2)
        dp = [[0] * (n + 1) for _ in range(m + 1)]
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if text1[i - 1] == text2[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1] + 1
                else:
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
        return dp[m][n]


class Solution_2:
    def longestCommonSubsequence(self, text1, text2):
        m, n = len(text1), len(text2)
        prev = [0] * (n + 1)
        for i in range(1, m + 1):
            cur = [0] * (n + 1)
            for j in range(1, n + 1):
                if text1[i - 1] == text2[j - 1]:
                    cur[j] = prev[j - 1] + 1
                else:
                    cur[j] = max(prev[j], cur[j - 1])
            prev = cur
        return prev[n]


if __name__ == "__main__":
    import time, tracemalloc
    cases = [("abcde", "ace", 3), ("abc", "abc", 3), ("abc", "def", 0),
             ("bsbininm", "jmjkbkjkv", 1)]
    tracemalloc.start()
    t0 = time.perf_counter()
    ok = True
    for a, b, want in cases:
        r1, r2 = Solution_1().longestCommonSubsequence(a, b), Solution_2().longestCommonSubsequence(a, b)
        ok &= r1 == r2 == want
        print(f'"{a}" x "{b}" -> A1={r1} A2={r2} (want {want}) {"PASS" if r1 == want and r2 == want else "FAIL"}')
    dt = time.perf_counter() - t0
    mem = tracemalloc.get_traced_memory()[1] / 1024
    tracemalloc.stop()
    print(f"{'PASS : all cases' if ok else 'FAIL'} | time: {dt*1e6:.0f} us | mem: {mem:.0f} KB")

File: neetcode150/13_1d_dp/1143_longest_common_subsequence.cpp
/*
 * =====================================================================
 * LeetCode 1143 : Longest Common Subsequence                      (Medium)
 * https://leetcode.com/problems/longest-common-subsequence/
 * Category   : 1-D dynamic programming
 * ---------------------------------------------------------------------
 * PROBLEM
 *   Longest subsequence common to two strings (subsequence = delete
 *   any chars, keep order). Return its length.
 * ---------------------------------------------------------------------
 * INTUITION
 *   Classic 2-sequence DP: dp[i][j] = LCS of prefixes text1[0..i),
 *   text2[0..j). Match → +1 along the diagonal; mismatch → take the
 *   max of dropping either tail. The table only needs the previous
 *   row, so memory collapses to O(min(m, n)).
 * ---------------------------------------------------------------------
 * APPROACH 1 · O(m·n) table (best)
 *   1. dp[i][j] = dp[i-1][j-1]+1 if chars equal (use the diagonal).
 *   2. Else max(dp[i-1][j], dp[i][j-1]).
 * APPROACH 2 · Rolling two rows (memory-optimized)
 *   Keep prev[] and cur[]; same recurrences, O(n) memory.
 * ---------------------------------------------------------------------
 * DEEP DIVE · Why the recurrence is "longest path in a grid"
 *   Think of matching as walking a grid: (i,j) is the state; equal
 *   chars are a diagonal step worth +1, unequal chars move you right
 *   or down for free. dp then is exactly the longest-path value, and
 *   the optimal substructure (prefix of an optimal solution is
 *   optimal) holds — the foundation of the recurrence. With both
 *   strings 1000 chars, the O(n²) memory table is 1M ints ≈ 8 MB;
 *   the rolling row drops that to 8 KB.
 * ---------------------------------------------------------------------
 * TIME COMPLEXITY : O(m·n).
 * MEMORY COMPLEXITY: O(m·n) or O(n) rolling.
 * =====================================================================
 */
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <chrono>
#include <sys/resource.h>

class Solution_1 {
public:
    int longestCommonSubsequence(std::string text1, std::string text2) {
        int m = static_cast<int>(text1.size());
        int n = static_cast<int>(text2.size());
        std::vector<std::vector<int>> dp(m + 1, std::vector<int>(n + 1, 0));
        for (int i = 1; i <= m; ++i)
            for (int j = 1; j <= n; ++j)
                dp[i][j] = (text1[i - 1] == text2[j - 1])
                               ? dp[i - 1][j - 1] + 1
                               : std::max(dp[i - 1][j], dp[i][j - 1]);
        return dp[m][n];
    }
};

class Solution_2 {
public:
    int longestCommonSubsequence(std::string text1, std::string text2) {
        int m = static_cast<int>(text1.size());
        int n = static_cast<int>(text2.size());
        std::vector<int> prev(n + 1, 0), cur(n + 1, 0);
        for (int i = 1; i <= m; ++i) {
            for (int j = 1; j <= n; ++j)
                cur[j] = (text1[i - 1] == text2[j - 1])
                             ? prev[j - 1] + 1
                             : std::max(prev[j], cur[j - 1]);
            std::swap(prev, cur);
        }
        return prev[n];
    }
};

static long long nowUs() {
    return std::chrono::duration_cast<std::chrono::microseconds>(
        std::chrono::steady_clock::now().time_since_epoch()).count();
}
static long memKb() { struct rusage r{}; getrusage(RUSAGE_SELF, &r); return r.ru_maxrss; }

int main() {
    struct Case { std::string a, b; int want; };
    const Case cases[] = {
        {"abcde", "ace", 3},
        {"abc", "abc", 3},
        {"abc", "def", 0},
        {"bsbininm", "jmjkbkjkv", 1}
    };
    const long mem0 = memKb();
    auto t0 = nowUs();
    bool allOk = true;
    for (const auto& c : cases) {
        int r1 = Solution_1().longestCommonSubsequence(c.a, c.b);
        int r2 = Solution_2().longestCommonSubsequence(c.a, c.b);
        bool ok = r1 == c.want && r2 == c.want;
        allOk = allOk && ok;
        std::cout << "\"" << c.a << "\" × \"" << c.b << "\" → A1=" << r1
                  << " A2=" << r2 << " (want " << c.want << ") "
                  << (ok ? "PASS" : "FAIL") << "\n";
    }
    auto t1 = nowUs();
    std::cout << (allOk ? "PASS : all cases" : "FAIL") << " | time: " << (t1 - t0)
              << " us | mem: " << (memKb() - mem0) << " KB\n";
    return 0;
}
